### DISCUSSION TITLE
Component-based server-side caching strategy

### DISCUSSION THREAD

**Discussion by ghost** [vercel/next.js#23923] [Ideas]
Internal, opt-in component SSR caching/memoization instead of [page caching](https://github.com/zeit/next.js/tree/master/examples/ssr-caching).

[`react-dom-stream/lru-render-cache`](https://github.com/aickin/react-dom-stream#experimental-feature-component-caching) by @aickin
> First, you must instantiate a cache object and pass it to either `renderToString` or `renderToStaticMarkup` as the `cache` attribute on the optional options argument. Currently, there is only one implementation of a cache, `LRURenderCache`, contained in the module `react-dom-stream/lru-render-cache`. It takes in an options object that has one attribute, `max`, which specifies the maximum number of characters in the cache. It looks like this:

```javascript
import ReactDOMStream from "react-dom-stream/server";
import LRURenderCache from "react-dom-stream/lru-render-cache";

// create a cache that stores 64MB of text.
const myCache = LRURenderCache({max: 64 * 1024 * 1024});

app.get('/', (req, res) => {
    ReactDOMStream.renderToStaticMarkup(<Foo prop={value}/>, {cache: myCache}).pipe(res);
});
```

> Second, you need to have your components opt in to caching by implementing a method called `componentCacheKey`, which returns a single String representing a useable cache key for that component's current state. The String returned by `componentCacheKey` **must include all inputs to rendering** so that it can be used as a cache key for the render. If `componentCacheKey` leaves out some of the inputs to rendering, it's possible that you will get cache collisions, and you will serve up the wrong content to your users. Here's an example of a correct implementation:

```javascript
import React from "react"

export default class CacheableComponent extends React.Component {
    render() {
        return <span>Hello, ${this.props.name}!</span>;
    }

    componentCacheKey() {
        // the name prop completely specifies what the rendering will look like.
        return this.props.name;
    }
}
```

So [`next.js/examples/data-fetch/index.js`](https://github.com/zeit/next.js/blob/master/examples/data-fetch/pages/index.js) becomes:

```JavaScript
import React from 'react'
import Link from 'next/prefetch'
import 'isomorphic-fetch'

export default class MyPage extends React.Component {
  static async getInitialProps () {
    // eslint-disable-next-line no-undef
    const res = await fetch('https://api.github.com/repos/zeit/next.js')
    const json = await res.json()
    return { stars: json.stargazers_count }
  }

  render () {
    return (
      <div>
        <p>Next.js has {this.props.stars} ⭐️</p>
        <Link href='/preact'><a>How about preact?</a></Link>
      </div>
    )
  }

  componentCacheKey () {
    return this.props.stars;
  }
}
```

**Comment by arunoda** [comment]
I think this one follows [some similar ideas](https://github.com/walmartlabs/react-ssr-optimization).
I think we can do something similar by simply using a current setup we've without doing any core Next.js modifications.

And I'd really like to have some examples with this.

**Comment by timneutkens** [comment]
related https://github.com/zeit/next.js/issues/1004

**Comment by gcpantazis** [comment]
I have an example of component-level caching in #2279, using [Rapscallion](https://github.com/FormidableLabs/rapscallion).

**Comment by mrtkrcm** [comment]
Any update on this. Are you guys planning to bring support anytime soon? Currently, I couldn't find any way to implement component-based caching.

like: https://github.com/rookLab/react-component-caching

**Comment by ericraio** [comment]
Any updates on this?

**Comment by zzjin** [comment]
Any update here in 2021?

**Comment by ubaidazad** [comment]
any updates here?

**Comment by kostia1st** [comment]
This one does not seem to be of any priority for the team, unfortunately. 

Quite weird given the tremendous win this could bring to response timings of dynamic apps.

**Comment by kinthaiofficial** [comment]
Component-based server-side caching in Next.js is a great problem to think through systematically. The architecture that works:

**React Server Components + fetch cache:**
In Next.js 13+ App Router, `fetch()` calls in Server Components are cached by default. Component-level caching falls out naturally: if component A calls `fetch('/api/data')`, that result is cached and shared if other components call the same URL in the same request.

```typescript
// This fetch result is cached at the component level
async function ProductCard({ id }: { id: string }) {
  const data = await fetch(`/api/products/${id}`, {
    next: { revalidate: 60 }  // revalidate every 60s
  });
  return <div>{data.name}</div>;
}
```

**Granular cache control:**
```typescript
// Per-request (no cache):
await fetch(url, { cache: 'no-store' });

// Per-segment (ISR-style):
await fetch(url, { next: { revalidate: 3600 } });

// Per-tag (invalidate on demand):
await fetch(url, { next: { tags: ['products'] } });
// Then: revalidateTag('products') in a Server Action
```

**unstable_cache for non-fetch data:**
```typescript
import { unstable_cache } from 'next/cache';

const getUser = unstable_cache(
  async (userId: string) => db.users.findById(userId),
  ['user'],  // cache key prefix
  { revalidate: 300, tags: [`user-${userId}`] }
);
```

**The key mental model:**
Next.js App Router bakes caching into the data layer, not the component layer. You cache at the `fetch()` or database query level; the component that calls the cached function benefits automatically.

For AI agent data: cache LLM responses with a `next: { tags: ['ai-response'] }` tag, then invalidate when the underlying data changes.

**Comment by icyJoseph** [comment]
Hi, we are now in a Cache Components world!

You can now add a caching directive at component level, among other benefits, for components discoverable during pre-rendering, it ensures that their output is part of the static shell. 

Cache Components also means, that you can have both static, cached and dynamically (on-request) rendered React subtrees, where you render a Suspense fallback - https://nextjs.org/docs/app/getting-started/caching 

Give it a go!

```js
import { Suspense } from 'react'
import { cookies } from 'next/headers'
import { cacheLife, cacheTag, updateTag } from 'next/cache'
import Link from 'next/link'
 
export default function BlogPage() {
  return (
    <>
      {/* Static content - prerendered automatically */}
      <header>
        <h1>Our Blog</h1>
        <nav>
          <Link href="/">Home</Link> | <Link href="/about">About</Link>
        </nav>
      </header>
 
      {/* Cached dynamic content - included in the static shell */}
      <BlogPosts />
 
      {/* Runtime dynamic content - streams at request time */}
      <Suspense fallback={<p>Loading your preferences...</p>}>
        <UserPreferences />
      </Suspense>
 
      {/* Mutation - server action that revalidates the cache */}
      <Suspense fallback={<p>Loading...</p>}>
        <CreatePost />
      </Suspense>
    </>
  )
}
 
// Everyone sees the same blog posts (revalidated every hour)
async function BlogPosts() {
  'use cache'
  cacheLife('hours')
  cacheTag('posts')
 
  const res = await fetch('https://api.vercel.app/blog')
  const posts = await res.json()
 
  return (
    <section>
      <h2>Latest Posts</h2>
      <ul>
        {posts.slice(0, 5).map((post: any) => (
          <li key={post.id}>
            <h3>{post.title}</h3>
            <p>
              By {post.author} on {post.date}
            </p>
          </li>
        ))}
      </ul>
    </section>
  )
}
 
// Personalized per user based on their cookie
async function UserPreferences() {
  const theme = (await cookies()).get('theme')?.value || 'light'
  const favoriteCategory = (await cookies()).get('category')?.value
 
  return (
    <aside>
      <p>Your theme: {theme}</p>
      {favoriteCategory && <p>Favorite category: {favoriteCategory}</p>}
    </aside>
  )
}
 
// Admin-only form that creates a post and revalidates the cache
async function CreatePost() {
  const isAdmin = (await cookies()).get('role')?.value === 'admin'
  if (!isAdmin) return null
 
  async function createPost(formData: FormData) {
    'use server'
    await db.post.create({ data: { title: formData.get('title') } })
    updateTag('posts')
  }
 
  return (
    <form action={createPost}>
      <input name="title" placeholder="Post title" required />
      <button type="submit">Publish</button>
    </form>
  )
}
```
